## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 516780`
Commit: `d185a52ae9b7ea519d0af1de4969d979a15830b1`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>rng-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>botanEsdm</li>
    <li>botanEsdm.bin</li>
    <li>botanEsdm.dev</li>
    <li>botanEsdm.doc</li>
    <li>botanEsdm.man</li>
    <li>esdm</li>
    <li>jitterentropy</li>
    <li>jitterentropy-rngd</li>
    <li>jitterentropy.dev</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-linux`</summary>
<details>
<summary>rng-tools</summary>
<pre>rngd_jitter.c:97:32: error: invalid use of undefined type &#x27;struct jent_notime_ctx&#x27;
   97 |         pthread_join(thread_ctx-&gt;notime_thread_id, NULL);
      |                                ^~
rngd_jitter.c:98:41: error: invalid use of undefined type &#x27;struct jent_notime_ctx&#x27;
   98 |         pthread_attr_destroy(&amp;thread_ctx-&gt;notime_pthread_attr);
      |                                         ^~
rngd_jitter.c: At top level:
rngd_jitter.c:122:30: error: initialization of &#x27;int (*)(void *, int (*)(void *), void *)&#x27; from incompatible pointer type &#x27;int (*)(void *, void * (*)(void *), void *)&#x27; [8;;https://gcc.gnu.org/onlinedocs/gcc-15.2.0/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Wincompatible-pointer-types8;;]
  122 |         .jent_notime_start = rngd_notime_start,
      |                              ^~~~~~~~~~~~~~~~~
rngd_jitter.c:122:30: note: (near initialization for &#x27;rngd_notime_thread_builtin.jent_notime_start&#x27;)
rngd_jitter.c:44:12: note: &#x27;rngd_notime_start&#x27; declared here
   44 | static int rngd_notime_start(void *ctx,
      |            ^~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:877: rngd-rngd_jitter.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory &#x27;/build/source&#x27;
make[1]: *** [Makefile:1035: all-recursive] Error 1
make[1]: Leaving directory &#x27;/build/source&#x27;
make: *** [Makefile:510: all] Error 2</pre>
</details>
</details>

--- a/nixos/modules/services/security/jitterentropy-rngd.nix
+++ b/nixos/modules/services/security/jitterentropy-rngd.nix
@@ -11,12 +11,53 @@ in
   options.services.jitterentropy-rngd = {
     enable = lib.mkEnableOption "jitterentropy-rngd service configuration";
     package = lib.mkPackageOption pkgs "jitterentropy-rngd" { };
+    osr = lib.mkOption {
+      type = lib.types.ints.between 3 20;
+      default = 3;
+      description = "Oversampling rate for jitterentropy (3 to 20)";
+    };
+    flags = lib.mkOption {
+      type = lib.types.int;
+      default = 0;
+      description = "Additional flags to pass to jitterentropy";
+    };
+    forceSP800-90B = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = "Force SP800-90B mode for entropy reading";
+    };
+    verbose = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = "Enable verbose log messages";
+    };
   };
 
-  config = lib.mkIf cfg.enable {
-    systemd.packages = [ cfg.package ];
-    systemd.services."jitterentropy".wantedBy = [ "basic.target" ];
-  };
+  config =
+    let
+      # use identical arguments for status and service execution,
+      # in order to get meaningful output
+      args =
+        "--osr ${builtins.toString cfg.osr} --flags ${builtins.toString cfg.flags}"
+        + lib.optionalString cfg.forceSP800-90B " --sp800-90b"
+        + lib.optionalString cfg.verbose " -vvv";
+    in
+    lib.mkIf cfg.enable {
+      systemd.packages = [ cfg.package ];
+      systemd.services."jitterentropy".wantedBy = [ "basic.target" ];
+      systemd.services."jitterentropy".serviceConfig = {
+        # logs used configuration for comparison
+        ExecStartPre = [
+          "-${cfg.package}/bin/jitterentropy-rngd --status ${args}"
+        ];
+        ExecStart = [
+          # clear old setting from built-in service file
+          ""
+          # use service from package with our configured args
+          "${cfg.package}/bin/jitterentropy-rngd ${args}"
+        ];
+      };
+    };
 
   meta.maintainers = with lib.maintainers; [ thillux ];
 }

--- a/pkgs/by-name/ji/jitterentropy-rngd/package.nix
+++ b/pkgs/by-name/ji/jitterentropy-rngd/package.nix
@@ -6,13 +6,13 @@
 
 stdenv.mkDerivation (finalAttrs: {
   pname = "jitterentropy-rngd";
-  version = "1.2.8";
+  version = "1.3.1";
 
   src = fetchFromGitHub {
     owner = "smuellerDD";
     repo = "jitterentropy-rngd";
     rev = "v${finalAttrs.version}";
-    hash = "sha256-LDym636ss3B1G/vrqatu9g5vbVEeDX0JQcxZ/IxGeY0=";
+    hash = "sha256-iXpeN0PAPk8mcaNXwj6TlyK57NSFNOVs/XmEmUG1gIg=";
   };
 
   enableParallelBuilding = true;
@@ -25,6 +25,13 @@ stdenv.mkDerivation (finalAttrs: {
 
     runHook postInstall
   '';
+
+  # this package internally compiles without optimization by choice,
+  # as it introduces more execution time jitter, therefore disable fortify.
+  hardeningDisable = [
+    "fortify"
+    "fortify3"
+  ];
 
   meta = {
     description = "Random number generator, which injects entropy to the kernel";

--- a/pkgs/by-name/ji/jitterentropy/package.nix
+++ b/pkgs/by-name/ji/jitterentropy/package.nix
@@ -48,6 +48,7 @@ stdenv.mkDerivation (finalAttrs: {
     platforms = lib.platforms.linux ++ lib.platforms.darwin;
     maintainers = with lib.maintainers; [
       johnazoidberg
+      thillux
     ];
   };
 })

